expo/expo
# feat(cli): add --template option to run

# Why

When you run `expo run:android` command, `expo` checks the project root if the native project folder is there to build the project. If the project doesn't have native code, it starts `prebuild` process to clone and extract the prebuilt project files by using `expo-template-bare-minimum` template. Because the original `expo` template's native `android` project is built with the latest `react-native` package which does not support tvOS functionalities, the CLI updates the `package.json` file with `react-native@LATEST_VERSION` and put the native project folder/files accordingly. Then the project is breaks. (_You can find the output in `Test Plan` section of this description_)

# How
To be able to use `expo` with the `react-native-tvos` to build TV apps, I had to use a custom template for `prebuild` process. I've seen that there is `template` option but it wasn't parsed from the arguments. I have added that argument and created a new template called [`expo-template-bare-minimum-tvos`](https://github.com/abdullahceylan/expo-template-bare-minimum-tvos) to use it for the `prebuild` process that happens when you use `run` command.

Otherwise, the TV project is not working within the published/ejected/prebuilt APK.

# Test Plan
I have tested this with one of my TV apps. You can reproduce the issue by following the steps below:

- Download the snack from https://snack.expo.dev/@abdullahceylanwtc/expo-react-native-tvos-demo
- Open `package.json` and change your imports to import react-native as `"react-native": "npm:react-native-tvos@latest",`, so that this package is used instead of the core react-native package.
- `yarn install`
- `yarn android`
- Observe that the `"react-native": "npm:react-native-tvos@latest",` is replaced with `"react-native": "0.69.5"` by `expo`
- If you continue, it will be built without an issue but the TV-related events like `onFocus`, `onBlur` won't work (when you move between the items, should see the effect) on Android TV emulator because they are removed from the core `react-native` package.
- Now, open your `package.json` file and revert the `react-native` import back to `"react-native": "npm:react-native-tvos@latest",`
- Then run `yarn android` again.
- Then you'll see the error below:

```
> Task :app:compileDebugJavaWithJavac FAILED
PROJECT_PATH/android/app/src/main/java/com/abdullahceylan/acnativevideo/newarchitecture/MainApplicationReactNativeHost.java:109: error: cannot find symbol
                    ReactNativeConfig.DEFAULT_CONFIG,
                                     ^
  symbol:   variable DEFAULT_CONFIG
  location: interface ReactNativeConfig
PROJECT_PATH/android/app/src/main/java/com/abdullahceylan/acnativevideo/MainActivity.java:76: error: method does not override or implement a method from a supertype
    @Override
    ^
Note: PROJECT_PATH/android/app/src/debug/java/com/abdullahceylan/acnativevideo/ReactNativeFlipper.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
2 errors

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.

BUILD FAILED in 40s
508 actionable tasks: 10 executed, 498 from cache
Error: PROJECT_PATH/android/gradlew exited with non-zero code: 1
Error: PROJECT_PATH/android/gradlew exited with non-zero code: 1
    at ChildProcess.completionListener (PROJECT_PATH/node_modules/@expo/spawn-async/build/spawnAsync.js:52:23)
    at Object.onceWrapper (node:events:646:26)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)
    ...
    at Object.spawnAsync [as default] (PROJECT_PATH/node_modules/@expo/spawn-async/build/spawnAsync.js:17:21)
    at spawnGradleAsync (PROJECT_PATH/node_modules/@expo/cli/build/src/start/platforms/android/gradle.js:72:46)
    at Object.assembleAsync (PROJECT_PATH/node_modules/@expo/cli/build/src/start/platforms/android/gradle.js:52:18)
    at runAndroidAsync (PROJECT_PATH/node_modules/@expo/cli/build/src/run/android/runAndroidAsync.js:31:24)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

So how I tested the fix is, I have created a new template based on `bare-minimum` template. Then I build the CLI and defined it as `nexpo` alias. Then in the project root:

- `rm -rf android` to remove the `android` folder previously extracted from the `expo-template-bare-minimum`
- Run `nexpo run:android --template https://github.com/abdullahceylan/expo-template-bare-minimum-tvos`
- Observe that the build is successful and `onFocus`, `onBlur` etc. work on the Android TV emulator properly.

# Checklist
